domain/operation/errors/errors.go

```diff
@@ -17,3 +21,28 @@ const (
 	// is queried and does not have a pending status.
 	TaskNotPending = errors.ConstError("task not pending")
 )
+
+// ActionNotDefined describes an error that occurs when the given charm does
+// not define the given action.
+type ActionNotDefined struct {
+	// CharmName is the name of the charm missing the action.
+	CharmName string
+	// UnitName is the name of the unit where the action has been requested.
+	UnitName string
+	// HasActions is true if the charm defines some actions.
+	HasActions bool
+}
+
+func (a ActionNotDefined) Error() string {
+	return fmt.Sprintf("action not defined for charm %q", a.CharmName)
+}
+
+func (a ActionNotDefined) Is(target error) bool {
```

Comment:
You shouldn't need to implement this. We have `HasType` helper in the error package which is the correct way to check if it exists in a chain. Suggestion would be to delete this.